lib/resumption/handover_unix.go

```diff
@@ -0,0 +1,173 @@
+//go:build unix
+
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+const sockSuffix = ".sock"
+
+func sockPath(dataDir string, token resumptionToken) string {
+	hash := sha256.Sum256(token[:])
+	// unix domain sockets are limited to 108 or 104 characters, so the full
+	// sha256 hash is a bit too much (64 bytes in hex or 44 in b64); truncating
+	// the hash to 128 bits still gives us more than enough headroom to just
+	// assume that we'll have no collisions (a probability of one in a
+	// quintillion with 26 billion concurrent connections)
+	return filepath.Join(dataDir, "handover", base64.RawURLEncoding.EncodeToString(hash[:16])+sockSuffix)
+}
+
+func sockDir(dataDir string) string {
+	return filepath.Join(dataDir, "handover")
+}
+
+var errNoDataDir error = &trace.NotFoundError{Message: "data dir not configured"}
+
+func (r *SSHServerWrapper) listenHandover(token resumptionToken) (net.Listener, error) {
+	if r.dataDir == "" {
+		return nil, trace.Wrap(errNoDataDir)
+	}
+
+	_ = os.MkdirAll(sockDir(r.dataDir), 0o700)
+	l, err := net.Listen("unix", sockPath(r.dataDir, token))
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	return l, nil
+}
+
+func (r *SSHServerWrapper) dialHandover(token resumptionToken) (net.Conn, error) {
+	if r.dataDir == "" {
+		return nil, trace.Wrap(errNoDataDir)
+	}
+
+	c, err := net.DialTimeout("unix", sockPath(r.dataDir, token), time.Second)
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	return c, nil
+}
+
+func filterNonConnectableSockets(ctx context.Context, paths []string) (filtered []string, lastErr error) {
+	filtered = paths[:0]
+
+	var d net.Dialer
+	for _, path := range paths {
+		c, err := d.DialContext(ctx, "unix", path)
+		if err == nil {
+			_ = c.Close()
+			continue
+		}
+
+		if errors.Is(err, os.ErrNotExist) {
+			continue
+		}
+
+		if errors.Is(err, syscall.ECONNREFUSED) {
+			filtered = append(filtered, path)
+			continue
+		}
+
+		lastErr = trace.ConvertSystemError(err)
+	}
+
+	return filtered, lastErr
```

Comment:
That reasoning doesn't really make sense to me, if we want to identify potential issues during cleanup then we'd want to see all the errors in case there were multiple different ones. 

If you're adamant that this is the correct approach then I won't fight you further, but you should at least add a comment explaining why you're doing something non-standard.

Reply:
I've capped the amount of errors when aggregating in f434358aac6303015e71e3644c3e5cd75d029d95, in both retainNonConnectableSockets and handoverCleanup. WDYT?